Rename + deprecate a bunch of methods for API consistency

fralalonde opened the pull request:
Also, `Cache*` constructors (named `wrap()`) were made private since they are not expected to be used directly but through fluent `.cached()` method.

All renamed methods were kept as `deprecated` for compatibility. 
List of renamed methods:
```
Prefixed
 - add_prefix(n) -> named(n)

AtomicBucket
 - set_drain(d) -> drain(d)
 - set_stats(s) -> stats(s)

MultiInput
 - input() -> new()  
 - add_target(t) -> target(t)

MultiInputScope
 - add_target(t) -> target(t)

MultiOutput
 - output() -> new()  
 - add_target(t) -> target(t)

MultiOutputScope
 - add_target(t) -> target(t)

Proxy
- set_target(t) -> target(t)
- set_default_target(t) -> default_target(t)

Input
- input() -> metrics()

Output 
- output() -> new_scope()
```

fralalonde commented:
Fixes for #36 .

To be discussed.

fralalonde commented:
@mixalturek what do you think of these changes?

fralalonde commented:
In addition to renaming methods, some concepts should just be documented better. Documenting things properly is _hard_, because it requires virtually un-learning in order to re-learn in order to develop proper explanations. The handbook, external examples and rustdoc could all take some work in that regard.

mixalturek commented on examples/multi_output.rs:
```diff
@@ -7,16 +7,16 @@ use std::time::Duration;
+    let same_type_metrics = MultiOutput::new()
+        .target(Stream::to_stderr().named("yeah"))
+        .target(Stream::to_stderr().named("ouch"))
+        .named("out_both").metrics();
```
"cool" in comment comes from `input`, `out_both` is here.

fralalonde replied:
Comment vs. code fixed.
See my answer to `Cache` above for the case of duplicated `Multi`.

fralalonde commented on HANDBOOK.md:
````diff
@@ -187,7 +187,7 @@ Some outputs such as statsd also have the ability to sample metrics.
 If enabled, sampling is done using pcg32, a fast random algorithm with reasonable entropy.
 
 ```$rust,skt-run,no_run
-let _app_metrics = Statsd::send_to("server:8125")?.sampled(Sampling::Random(0.01)).input();
````
I had this planned at some point but thought it was not worth the added complexity. Could you open a separate issue to track it?

fralalonde commented on examples/clopwizard.rs:
```diff
@@ -26,15 +26,15 @@ fn main() {
-    AtomicBucket::set_default_stats(stats_all);
+    Proxy::default().target(all_buckets);
+    AtomicBucket::default_drain(Stream::to_stdout());
+    AtomicBucket::default_stats(stats_all);
```
Yes, the defaults set global share state. So if you multiple Buckets, you can configure them all at once. Then maybe override the defaults for just one of the buckets using a special set of `stats` or a different `target`.

fralalonde commented on src/cache/cache_out.rs:
```diff
@@ -12,10 +12,12 @@ use std::sync::{Arc, RwLock};
-/// This is useless if all metrics are statically declared but can provide performance
-/// benefits if some metrics are dynamically defined at runtime.
+/// This can provide performance benefits for metrics that are dynamically defined at runtime.
+/// Caching is useless if all metrics are statically declared.
```
They are identical except for the guarantees they provide, as `Output`s are not thread safe while `Input`s are. I'm thinking that maybe output caching could be dropped for simplicty, and keep it only on input, closest to the metrics declaring code.

`Multi` are duplicated the same way, except that there could be instances where you want the dispatching to be performed from the same thread (thus on the `Output` side), for example when using `async` queue. 

I would prefer these API splits to be generic but could not come up with a way to it, instead in the case of Cache the API hides the duplication by offering same-named `cached()` method both on `Input` and `Outptut`.

fralalonde commented on src/core/locking.rs:
```diff
@@ -30,11 +30,11 @@ impl InputScope for LockingOutput {
         let name = self.prefix_append(name);
         // lock when creating metrics
-        let raw_metric = self.inner.lock().expect("OutputScope Lock").new_metric(name, kind);
+        let raw_metric = self.inner.lock().expect("LockingOutput").new_metric(name, kind);
```
From memory, I adopted that convention of terseness after reading the messages produced by failing `expect()`s which already provides some context. I also dont want to polute code with extra text... Actually these should all be defined as constants, then the text could be longer if required.

fralalonde commented on src/core/mod.rs:
```diff
@@ -30,7 +30,7 @@ pub mod test {
     #[test]
     fn test_to_void() {
-        let c = void::Void::metrics().input();
+        let c = void::Void::metrics().metrics();
```
LOL yes, this is horrible...

fralalonde commented on src/multi/multi_out.rs:
```diff
@@ -30,16 +30,30 @@ impl Output for MultiOutput {
+
+    /// Add a target to the dispatch list.
+    /// Returns a clone of the original object.
+    pub fn target<OUT: Output + Send + Sync + 'static>(&self, out: OUT) -> Self {
```
I wondered about this too, but went all the way for consistency with the other changes. I think sticking with `add_target` would also imply using `add_name` instead of `named` on `Attributes`, because both method essentially act the same way, cloning & appending the element to an internal list.

fralalonde commented:
PR updated with suggestions and latest `master` changes. Waiting on input on the mentioned `add_target` / `add_name` situation before merging and releasing 0.7.3

mixalturek commented:
I don't have any strong opinion about target/add_target, it should be ok to let it as is.

fralalonde commented:
Ok, last call for this one. 

First,  I've reverted to `add_target()` for `Multi`.

Second, I introduced `add_name()` in addition to `named()`. The big difference is that `named()` now _replaces_ any names currently defined on the component, while `add_name` will append the name to the existing ones (as `named()` did before). For a single-name component, both methods are equivalent. I think this is the most logical arrangement, both semantically and ergonomically.
